reinforcement-learning-games/zombsole-gym-example/dqn/q_learning.py

```diff
@@ -0,0 +1,184 @@
+        self.actions = range(0, env.action_space.n) # TODO: Improve this
+        self.feedback_size = env.get_frame_size()
+        print("feedback_size: %s" % (self.feedback_size,))
+        self.callback = callback
+        self.summary_writer = summary_writer
+        
+        self.config = config
+        self.batch_size = config['batch_size']
+        self.n_episode = config['num_episode']
+        self.capacity = config['capacity']
+        self.epsilon_decay = config['epsilon_decay']
+        self.epsilon_min = config['epsilon_min']
+        self.num_frames = config['num_frames']
+        self.num_nullops = config['num_nullops']
+        self.time_between_two_copies = config['time_between_two_copies']
+        self.input_scale = config['input_scale']
+        self.update_interval = config['update_interval']
+        self.directory = directory
+
+        self.verbose = verbose
+
+        self._init_modules()
+        
+    def _init_modules(self):
+        
+        # Replay memory
+        self.replay_memory = ReplayMemory(history_len=self.num_frames, 
+                                          capacity=self.capacity, 
+                                          batch_size=self.batch_size,
+                                          input_scale=self.input_scale)
+        
+        input_shape = self.feedback_size + (self.num_frames,)
+        # Q-network
+        self.q_network = QNetwork(input_shape=input_shape, n_outputs=len(self.actions), 
+                                  network_type=self.config['network_type'], scope='q_network')
+        # Target network
+        self.target_network = QNetwork(input_shape=input_shape, n_outputs=len(self.actions), 
+                                       network_type=self.config['network_type'], scope='target_network')
+        # Optimizer
+        self.optimizer = Optimizer(config=self.config, 
+                                   feedback_size=self.feedback_size, 
+                                   q_network=self.q_network, 
+                                   target_network=self.target_network, 
+                                   replay_memory=self.replay_memory)
+        # Ops for updating target network
+        self.clone_op = self.target_network.get_clone_op(self.q_network)
+        # For tensorboard
+        self.t_score = tf.placeholder(dtype=tf.float32, shape=[], name='new_score')
+        tf.summary.scalar("score", self.t_score, collections=['dqn'])
+        self.summary_op = tf.summary.merge_all('dqn')
+    
+    def set_summary_writer(self, summary_writer=None):
+        self.summary_writer = summary_writer
+        self.optimizer.set_summary_writer(summary_writer)
+    
+    def choose_action(self, sess, state, epsilon_greedy):
+        if numpy.random.binomial(1, epsilon_greedy) == 1:
+            action = random.choice(self.actions)
+        else:
+            x = numpy.asarray(numpy.expand_dims(state, axis=0) / self.input_scale, dtype=numpy.float32)
```

Comment:
Note that rescaling is done here.  It is worth noting that calls to `self.replay_memory.phi` (and `_phi`) do not rescale, but `self.replay_memory.sample()` does.